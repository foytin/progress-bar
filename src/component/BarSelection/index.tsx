import React from "react";
import PropTypes from "prop-types";

interface BarSelectionProps {
  bars: Array<number>;
  onChange: (arg: string) => void;
}

const BarSelection = (props: BarSelectionProps): React.ReactElement => {
  const { bars, onChange } = props;

  return (
    <select
      className="custom-select custom-select-lg"
      onBlur={(e) => onChange(e.target.value)}
      data-testid="bar-selection"
    >
      {bars.map((_bar, index) => (
        <option key={`progess${index}`} value={index}>
          #Progress {index + 1}
        </option>
      ))}
    </select>
  );
};

BarSelection.propTypes = {
  bars: PropTypes.arrayOf(PropTypes.number).isRequired,
  onChange: PropTypes.func.isRequired,
};

export default BarSelection;
