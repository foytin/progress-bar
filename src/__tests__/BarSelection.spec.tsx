import React from "react";
import { render, fireEvent } from "@testing-library/react";

import BarSelection from "../component/BarSelection";

describe("BarSelection component", () => {
  it("Correctly Render", () => {
    const wrapper = render(
      <BarSelection bars={[23, 66, 58]} onChange={jest.fn()} />
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it("if bars has 3 records, should have 3 options", () => {
    const wrapper = render(
      <BarSelection bars={[23, 66, 58]} onChange={jest.fn()} />
    );
    expect(wrapper.getAllByRole("option")).toHaveLength(3);
  });

  it("Second option value should be 1, and text should be #Progress 2", () => {
    const wrapper = render(
      <BarSelection bars={[23, 66, 58]} onChange={jest.fn()} />
    );
    expect(wrapper.getAllByRole("option")[1]).toHaveTextContent("#Progress 2");
    expect(wrapper.getAllByRole("option")[1]).toHaveValue("1");
  });

  it("when select on option, mock fn will be fired along with option value", () => {
    const fn = jest.fn();
    const value = "2";
    const wrapper = render(<BarSelection bars={[23, 66, 58]} onChange={fn} />);
    fireEvent.blur(wrapper.getByTestId("bar-selection"), { target: { value } });
    expect(fn).toHaveBeenLastCalledWith(value);
  });
});
