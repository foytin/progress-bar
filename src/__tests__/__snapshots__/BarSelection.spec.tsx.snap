// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarSelection component Correctly Render 1`] = `
<div>
  <select
    class="custom-select custom-select-lg"
    data-testid="bar-selection"
  >
    <option
      value="0"
    >
      #Progress 1
    </option>
    <option
      value="1"
    >
      #Progress 2
    </option>
    <option
      value="2"
    >
      #Progress 3
    </option>
  </select>
</div>
`;
